File: frontend/src/lib/money.test.ts
import { describe, it, expect } from 'vitest';
import { formatMoney, parseMoney, moneySpoken } from './money';

describe('formatMoney', () => {
  it('renders full (Vietnamese norm with dot thousands + đ suffix)', () => {
    expect(formatMoney(0)).toBe('0đ');
    expect(formatMoney(1_000)).toBe('1.000đ');
    expect(formatMoney(120_000)).toBe('120.000đ');
    expect(formatMoney(1_234_567)).toBe('1.234.567đ');
  });

  it('renders compact with k / tr buckets, no trailing .0', () => {
    expect(formatMoney(500, 'compact')).toBe('500');
    expect(formatMoney(1_000, 'compact')).toBe('1k');
    expect(formatMoney(1_200, 'compact')).toBe('1.2k');
    expect(formatMoney(120_000, 'compact')).toBe('120k');
    expect(formatMoney(1_000_000, 'compact')).toBe('1tr');
    expect(formatMoney(1_200_000, 'compact')).toBe('1.2tr');
    expect(formatMoney(2_000_000, 'compact')).toBe('2tr');
    expect(formatMoney(1_234_567, 'compact')).toBe('1.2tr');
  });

  it('handles non-finite / negative defensively', () => {
    expect(formatMoney(NaN)).toBe('');
    expect(formatMoney(-120_000)).toBe('-120.000đ');
  });
});

describe('parseMoney', () => {
  it('parses plain integers', () => {
    expect(parseMoney('120000')).toBe(120_000);
    expect(parseMoney('0')).toBe(0);
  });

  it('parses dot-thousands (Vietnamese norm)', () => {
    expect(parseMoney('120.000')).toBe(120_000);
    expect(parseMoney('1.234.567')).toBe(1_234_567);
  });

  it('parses comma-thousands (US convention, tolerated)', () => {
    expect(parseMoney('120,000')).toBe(120_000);
  });

  it('parses compact k / nghìn forms', () => {
    expect(parseMoney('120k')).toBe(120_000);
    expect(parseMoney('120 nghìn')).toBe(120_000);
    expect(parseMoney('120nghin')).toBe(120_000);
    expect(parseMoney('1.5k')).toBe(1_500);
  });

  it('parses compact tr / triệu forms', () => {
    expect(parseMoney('1tr')).toBe(1_000_000);
    expect(parseMoney('1.5tr')).toBe(1_500_000);
    expect(parseMoney('1.5 triệu')).toBe(1_500_000);
    expect(parseMoney('1.2tr')).toBe(1_200_000);
  });

  it('strips trailing đ / đồng / vnđ', () => {
    expect(parseMoney('120.000đ')).toBe(120_000);
    expect(parseMoney('120 nghìn đồng')).toBe(120_000);
    expect(parseMoney('120000 vnđ')).toBe(120_000);
  });

  it('is whitespace + case tolerant', () => {
    expect(parseMoney('  120 K  ')).toBe(120_000);
    expect(parseMoney('1.2 TR')).toBe(1_200_000);
  });

  it('throws on empty / garbage / negative', () => {
    expect(() => parseMoney('')).toThrow();
    expect(() => parseMoney('   ')).toThrow();
    expect(() => parseMoney('abc')).toThrow();
    expect(() => parseMoney('-1000')).toThrow();
  });
});

describe('round-trip (compact → parse)', () => {
  it('preserves round amounts exactly', () => {
    for (const n of [1_000, 120_000, 1_000_000, 1_200_000, 2_000_000]) {
      expect(parseMoney(formatMoney(n, 'compact'))).toBe(n);
    }
  });

  it('rounds 1234567 down to 1.2tr precision (compact is lossy by design)', () => {
    const formatted = formatMoney(1_234_567, 'compact');
    expect(formatted).toBe('1.2tr');
    expect(parseMoney(formatted)).toBe(1_200_000);
  });
});

describe('round-trip (full → parse)', () => {
  it('preserves any amount exactly', () => {
    for (const n of [0, 1_000, 120_000, 1_234_567, 999_999_999]) {
      expect(parseMoney(formatMoney(n))).toBe(n);
    }
  });
});

describe('moneySpoken', () => {
  it('matches AC examples from Story 1.1 Dev Notes', () => {
    expect(moneySpoken(120_000)).toBe('120 nghìn đồng');
    expect(moneySpoken(1_234_567)).toBe('1 triệu 200 nghìn đồng');
  });

  it('drops the nghìn segment when amount is exact millions', () => {
    expect(moneySpoken(1_000_000)).toBe('1 triệu đồng');
    expect(moneySpoken(2_000_000)).toBe('2 triệu đồng');
  });

  it('renders fractional millions to one-tenth precision', () => {
    expect(moneySpoken(1_500_000)).toBe('1 triệu 500 nghìn đồng');
    expect(moneySpoken(2_500_000)).toBe('2 triệu 500 nghìn đồng');
  });

  it('handles sub-thousand and zero', () => {
    expect(moneySpoken(0)).toBe('0 đồng');
    expect(moneySpoken(500)).toBe('500 đồng');
    expect(moneySpoken(999)).toBe('999 đồng');
  });

  it('rounds sub-million amounts to nearest thousand', () => {
    expect(moneySpoken(12_499)).toBe('12 nghìn đồng');
    expect(moneySpoken(12_500)).toBe('13 nghìn đồng');
  });

  it('returns empty string for invalid inputs', () => {
    expect(moneySpoken(NaN)).toBe('');
    expect(moneySpoken(-120_000)).toBe('');
  });
});

File: backend/internal/money/money_test.go
package money

import (
	"strings"
	"testing"
)

func TestFormat_Full(t *testing.T) {
	cases := []struct {
		in   int64
		want string
	}{
		{0, "0đ"},
		{1_000, "1.000đ"},
		{120_000, "120.000đ"},
		{1_234_567, "1.234.567đ"},
		{-120_000, "-120.000đ"},
	}
	for _, c := range cases {
		got := Format(c.in, Full)
		if got != c.want {
			t.Errorf("Format(%d, Full) = %q; want %q", c.in, got, c.want)
		}
	}
}

func TestFormat_Compact(t *testing.T) {
	cases := []struct {
		in   int64
		want string
	}{
		{500, "500"},
		{1_000, "1k"},
		{1_200, "1.2k"},
		{120_000, "120k"},
		{1_000_000, "1tr"},
		{1_200_000, "1.2tr"},
		{2_000_000, "2tr"},
		{1_234_567, "1.2tr"},
	}
	for _, c := range cases {
		got := Format(c.in, Compact)
		if got != c.want {
			t.Errorf("Format(%d, Compact) = %q; want %q", c.in, got, c.want)
		}
	}
}

func TestParse_Accepts(t *testing.T) {
	cases := []struct {
		in   string
		want int64
	}{
		// Plain integers.
		{"120000", 120_000},
		{"0", 0},

		// Dot-thousands (Vietnamese norm).
		{"120.000", 120_000},
		{"1.234.567", 1_234_567},

		// Comma-thousands (US convention).
		{"120,000", 120_000},

		// Compact k / nghìn.
		{"120k", 120_000},
		{"120 nghìn", 120_000},
		{"120nghin", 120_000},
		{"1.5k", 1_500},

		// Compact tr / triệu.
		{"1tr", 1_000_000},
		{"1.5tr", 1_500_000},
		{"1.5 triệu", 1_500_000},
		{"1.2tr", 1_200_000},

		// With currency suffix.
		{"120.000đ", 120_000},
		{"120 nghìn đồng", 120_000},
		{"120000 vnđ", 120_000},

		// Whitespace + case.
		{"  120 K  ", 120_000},
		{"1.2 TR", 1_200_000},
	}
	for _, c := range cases {
		got, err := Parse(c.in)
		if err != nil {
			t.Errorf("Parse(%q) unexpected error: %v", c.in, err)
			continue
		}
		if got != c.want {
			t.Errorf("Parse(%q) = %d; want %d", c.in, got, c.want)
		}
	}
}

func TestParse_Rejects(t *testing.T) {
	cases := []string{"", "   ", "abc", "-1000"}
	for _, in := range cases {
		_, err := Parse(in)
		if err == nil {
			t.Errorf("Parse(%q) expected error, got nil", in)
		}
	}
}

func TestRoundTrip_Compact(t *testing.T) {
	// Compact preserves exactly for round amounts.
	for _, n := range []int64{1_000, 120_000, 1_000_000, 1_200_000, 2_000_000} {
		formatted := Format(n, Compact)
		got, err := Parse(formatted)
		if err != nil {
			t.Errorf("Parse(Format(%d, Compact)) = error %v", n, err)
			continue
		}
		if got != n {
			t.Errorf("RoundTrip(%d) = %d; want %d", n, got, n)
		}
	}

	// 1_234_567 rounds to 1.2tr (compact is lossy by design).
	formatted := Format(1_234_567, Compact)
	if formatted != "1.2tr" {
		t.Errorf("Format(1234567, Compact) = %q; want %q", formatted, "1.2tr")
	}
	got, err := Parse(formatted)
	if err != nil {
		t.Fatalf("Parse(%q) error: %v", formatted, err)
	}
	if got != 1_200_000 {
		t.Errorf("Parse(%q) = %d; want %d", formatted, got, 1_200_000)
	}
}

func TestRoundTrip_Full(t *testing.T) {
	for _, n := range []int64{0, 1_000, 120_000, 1_234_567, 999_999_999} {
		formatted := Format(n, Full)
		got, err := Parse(formatted)
		if err != nil {
			t.Errorf("Parse(Format(%d, Full)) = error %v", n, err)
			continue
		}
		if got != n {
			t.Errorf("RoundTrip(%d) = %d; want %d", n, got, n)
		}
	}
}

func TestSpoken(t *testing.T) {
	cases := []struct {
		in   int64
		want string
	}{
		{0, "0 đồng"},
		{500, "500 đồng"},
		{999, "999 đồng"},
		{120_000, "120 nghìn đồng"},
		{12_500, "13 nghìn đồng"}, // rounds nearest
		{12_499, "12 nghìn đồng"},
		{1_000_000, "1 triệu đồng"},
		{2_000_000, "2 triệu đồng"},
		{1_500_000, "1 triệu 500 nghìn đồng"},
		{2_500_000, "2 triệu 500 nghìn đồng"},
		{1_234_567, "1 triệu 200 nghìn đồng"}, // rounds to 1.2tr precision
		{1_200_000, "1 triệu 200 nghìn đồng"},
	}
	for _, c := range cases {
		got := Spoken(c.in)
		if got != c.want {
			t.Errorf("Spoken(%d) = %q; want %q", c.in, got, c.want)
		}
	}

	// Negative returns empty string.
	if got := Spoken(-1); got != "" {
		t.Errorf("Spoken(-1) = %q; want empty string", got)
	}
}

func TestSpoken_ContainsUnits(t *testing.T) {
	// Sanity: every non-zero output ends with " đồng".
	for _, n := range []int64{500, 1_000, 120_000, 1_500_000} {
		got := Spoken(n)
		if !strings.HasSuffix(got, " đồng") {
			t.Errorf("Spoken(%d) = %q; expected trailing ' đồng'", n, got)
		}
	}
}
